File: codes/contest/lintcode/count-of-range-sum.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <vector>
using namespace std;

struct Node {
    int value;
    int height;
    int lenum;
    Node* left;
    Node* right;
    Node(int value) {
        this->value = value;
        height = 1;
        lenum = 1;
        left = NULL;
        right = NULL;
    }
};

int get_height(Node* root) {
    if (root == NULL) return 0;
    return root->height;
}

Node* left_rotate(Node* root) {
    Node* left = root->left;
    root->left = left->right;
    root->lenum = root->lenum - left->lenum;
    root->height = max(get_height(root->left), get_height(root->right)) + 1;
    left->right = root;
    left->height = max(get_height(left->left), get_height(left->right)) + 1;
    return left;
}

Node* right_rotate(Node* root) {
    Node* right = root->right;
    root->right = right->left;
    root->height = max(get_height(root->left), get_height(root->right)) + 1;
    right->lenum = right->lenum + root->lenum;
    right->left = root;
    right->height = max(get_height(right->left), get_height(right->right)) + 1;
    return right;
}

Node* balance(Node* root) {
    int lh = get_height(root->left);
    int rh = get_height(root->right);

    if ((lh - rh) >= 2) {
        int llh = get_height(root->left->left);
        int lrh = get_height(root->left->right);
        if (lrh > llh) {
            root->left = right_rotate(root->left);
        }
        root = left_rotate(root);
    } else if ((rh - lh) >= 2) {
        int rlh = get_height(root->right->left);
        int rrh = get_height(root->right->right);
        if (rlh > rrh) {
            root->right = left_rotate(root->right);
        }
        root = right_rotate(root);
    } else {
        root->height = max(lh, rh) + 1;
    }
    return root;
}

Node* insert_node(Node* root, int value) {
    if (root == NULL) {
        Node* node = new Node(value);
        return node;
    }

    if (root->value == value) {
        root->lenum += 1;
    } else if (root->value > value) {
        root->left = insert_node(root->left, value);
        root->lenum += 1;
    } else {
        root->right = insert_node(root->right, value);
    }
    root = balance(root);
    return root;
}

int query(Node* root, int value) {
    if (root == NULL) {
        return 0;
    }
    if (root->value == value) {
        return root->lenum;
    } else if (root->value > value) {
        return query(root->left, value);
    } else {
        int count = query(root->right, value);
        count += root->lenum;
        return count;
    }
}

class Solution {
   public:
    /**
     * @param nums: a list of integers
     * @param lower: a integer
     * @param upper: a integer
     * @return: return a integer
     */
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        // write your code here
        Node* root = NULL;

        int acc = 0;
        int res = 0;
        for (size_t i = 0; i < nums.size(); i++) {
            int num = nums[i];
            acc += num;
            int lo = query(root, acc - upper - 1);
            int hi = query(root, acc - lower);
            res += (hi - lo);
            root = insert_node(root, acc);
            if (acc >= lower && acc <= upper) {
                res += 1;
            }
        }
        return res;
    }
};

File: codes/contest/lintcode/count-of-range-sum.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        self.lenum = 1
        self.height = 1


def height(r):
    return r.height if r else 0


def lenum(r):
    return r.lenum if r else 0


def left_rotate(root):
    # print('left_rotate!')
    left = root.left
    root.left = left.right
    root.lenum = root.lenum - lenum(left)
    root.height = max(height(root.left), height(root.right)) + 1

    left.right = root
    left.height = max(height(left.left), height(left.right)) + 1
    return left


def right_rotate(root):
    # print('right_rotate!')
    right = root.right
    root.right = right.left
    root.height = max(height(root.left), height(root.right)) + 1

    right.lenum = right.lenum + lenum(root)
    right.left = root
    right.height = max(height(right.left), height(right.right)) + 1
    return right


def balance(root):
    lh = height(root.left)
    rh = height(root.right)

    if (lh - rh) >= 2:
        llh = height(root.left.left)
        lrh = height(root.left.right)
        if lrh > llh:
            root.left = right_rotate(root.left)
        root = left_rotate(root)

    elif (rh - lh) >= 2:
        rlh = height(root.right.left)
        rrh = height(root.right.right)
        if rlh > rrh:
            root.right = left_rotate(root.right)
        root = right_rotate(root)

    else:
        root.height = max(lh, rh) + 1

    return root


def insert(root, value, do_balance):
    if root is None:
        return Node(value)

    if root.value == value:
        root.lenum += 1
    elif root.value > value:
        root.left = insert(root.left, value, do_balance)
        root.lenum += 1
    else:
        root.right = insert(root.right, value, do_balance)
    if do_balance:
        root = balance(root)
    return root


def query(root, value):
    if root is None:
        return 0

    if root.value == value:
        return root.lenum
    elif root.value > value:
        return query(root.left, value)
    else:
        count = query(root.right, value)
        count += root.lenum
        return count


class Tree:
    def __init__(self, do_balance=True):
        self.root = None
        self.do_balance = do_balance

    def height(self):
        return height(self.root)

    def insert_value(self, value):
        self.root = insert(self.root, value, do_balance=self.do_balance)

    def query_value(self, value):
        return query(self.root, value)


class Solution:
    """
    @param nums: a list of integers
    @param lower: a integer
    @param upper: a integer
    @return: return a integer
    """

    def countRangeSum(self, nums, lower, upper):
        # write your code here
        tree = Tree()
        # dtree = Tree(do_balance=False)
        acc = 0
        res = 0
        for num in nums:
            # print(tree.height())
            acc += num
            lo = tree.query_value(acc - upper - 1)
            hi = tree.query_value(acc - lower)

            # dlo = dtree.query_value(acc - upper - 1)
            # dhi = dtree.query_value(acc - lower)
            # if dlo != lo or dhi != hi:
            #     print('mismatch!!')

            res += (hi - lo)
            # print('{} {} {} {}'.format(acc - upper - 1, acc - lower, lo, hi))
            tree.insert_value(acc)
            # dtree.insert_value(acc)
            if lower <= acc <= upper:
                res += 1
        return res
